tests/tests/compiler_docs.rs

```diff
@@ -0,0 +1,169 @@
+use rune::ast::Span;
+use rune::termcolor::{ColorChoice, StandardStream};
+use rune_tests::{sources};
+
+struct DocVisitor {
+    expected: BTreeMap<&'static str, Vec<&'static str>>,
+    collected: BTreeMap<String, Vec<String>>
+}
+
+impl CompileVisitor for DocVisitor {
+
+    fn visit_doc_comment(&mut self, _: SourceId, meta: MetaRef<'_>, _: Span, doc: &str) {
+        let key = meta.item.to_string();
+        let vec = if let Some(vec) = self.collected.get_mut(&key) {
+            vec
+        } else {
+            self.collected.insert(key.clone(), Vec::new());
+            self.collected.get_mut(&key).unwrap()
+        };
+
+        vec.push(doc.to_string());
+    }
+
+}
+
+impl DocVisitor {
+
+    fn assert(&self) {
+        for (k, expected) in &self.expected {
+            let against = if let Some(vec) = self.collected.get(*k) {
+                vec
+            } else {
+                panic!("missing documentation for item {k:?}");
+            };
+
+            let mut idx: usize = 0;
+            while idx < expected.len() {
+                if let Some(collected) = against.get(idx) {
+                    assert_eq!(collected, expected[idx], "mismatched docstring:\n\tgot: {:?}\n\texp: {:?}", collected, expected[idx]);
+                } else {
+                    panic!("missing docstrings, expected: {:?}", expected[idx]);
+                }
+
+                idx += 1;
+            }
+
+            if expected.len() < against.len() {
+                let (_, extras) = against.split_at(expected.len());
+                panic!("extra docstrings: {:?}", extras);
+            }
+        }
+
+        if self.collected.len() > self.expected.len() {
+            let vec = self.collected.keys().filter(|it| !self.expected.contains_key(it.as_str())).collect::<Vec<_>>();
+            panic!("encountered more documented items than expected: {vec:?}");
+        }
+    }
+
+}
+
+macro_rules! expect_docs {
+    ($($typename:literal => { $($docstr:literal)* })+) => {
+        {
+            let mut expected = BTreeMap::new();
+
+            $(
+            #[allow(unused_mut)]
+            let mut vec = Vec::new();
+            $(vec.push($docstr);
+            )*
+
+            expected.insert($typename, vec);
+            )+
+
+            DocVisitor {
+                expected,
+                collected: BTreeMap::new()
+            }
+        }
+    };
+}
+
+#[test]
+fn harvest_docs() {
+    let mut diagnostics = Diagnostics::new();
+    let mut vis = expect_docs! {
+        "stuff" => { " Top-level function.\n" }
+        "Struct" => {
+            " Top-level struct.\n"
+            " Second line!\n"
+        }
+        "Enum" => { "\n         * Top-level enum.\n         " }
+        "Enum::A" => { " Enum variant A.\n" }
+        "Enum::B" => { " Enum variant B.\n" }
+        "constant" => { " Top-level constant.\n" }
+
+        // Curiously, the module item itself never get registered. This always fails.
```

Comment:
See comment about files. Modules suffer from two issues that need to be addressed:

* We currently don't treat these as meta items (e.g. module information is not required for compilation) so we don't call the associated visitors.
* Interior doc comments require that we traverse the inside of the module and collect those comments somehow.

Both of these require minor changes. The easiest might be to simply make sure modules register `PrivMeta` items. This would even to provide [better diagnostics](https://rune-rs.github.io/play/?c=bW9kIHRlc3Qgewp9CgpwdWIgZm4gbWFpbigpIHsKICAgIHRlc3QoKQp9Cg%3D%3D) during compilation.

So instead of:
```
error: compile error
  ┌─ entry:5:5
  │
5 │     test()
  │     ^^^^ missing item `test`
```
We could have something like:
```
error: compile error
  ┌─ entry:5:5
  │
5 │     test()
  │     ^^^^ `test` is a module which cannot be called
```